File: python/src/problem/leetcode/medium/leetcode_22.py
# coding=utf-8
from typing import List
import sys

from src.data_structure.data_structure import ListNode


class Solution:
    """
    括号生成
    """

    def generate(self, res: List[str], s: str, left: int, right: int):
        """

        :param res:
        :param s:
        :param left:
        :param right:
        :return:
        """
        if left == 0 and right == 0:
            res.append(s)
        else:
            if left > 0:
                self.generate(res, s + "(", left - 1, right)
            if right > left:
                self.generate(res, s + ")", left, right - 1)

    def generate_parenthesis(self, n: int) -> List[str]:
        """
        Time: O(4^n / sqrt(n)), Space: O(n)      卡特兰数
        :param n:
        :return:
        """
        res = []
        if n <= 0:
            return res
        self.generate(res, "", n, n)
        return res

File: c/medium/leetcode_22.c
/**
 * Created by ywh on 10/08/2020.
 */
#include <stdlib.h>
#include <string.h>

#define MAX 5000

/**
 *
 * @param ret
 * @param returnSize
 * @param str
 * @param left
 * @param right
 */
void generate(char ** ret, int* returnSize, char * str, int left, int right) {
    if (left == 0 && right == 0) {
        strcpy(ret[(*returnSize)++], str);
        return;
    }
    char *nextStr = (char *) malloc(strlen(str) + 2);
    if (left > 0) {
        char ca[2] = {'('};
        strcpy(nextStr, str);
        strcat(nextStr, ca);
        generate(ret, returnSize, nextStr, left - 1, right);
    }
    if (right > left) {
        char ca[2] = {')'};
        strcpy(nextStr, str);
        strcat(nextStr, ca);
        generate(ret, returnSize, nextStr, left, right - 1);
    }
}

/**
 * 括号的合法排列
 *
 * Time: O(4^n / sqrt(n)), Space: O(n)
 *
 * @param n
 * @param returnSize
 * @return
 */
char ** generateParenthesis(int n, int* returnSize){
    *returnSize = 0;
    if (n <= 0) {
        return NULL;
    }
    char **ret = (char **) malloc(sizeof(char *) * MAX);
    for (int i = 0; i < MAX; ++i) {
        memset(ret[i] = malloc(n * 2 + 1), 0, n * 2 + 1);
    }
    generate(ret, returnSize, "", n, n);
    return ret;
}